library/aesce.c

```diff
@@ -0,0 +1,326 @@
+/*
+ *  Arm64 crypto engine support functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <string.h>
+#include "common.h"
+
+#if defined(MBEDTLS_AESCE_C)
+
+#include "aesce.h"
+
+#if defined(MBEDTLS_HAVE_ARM64)
+
+#if !defined(__ARM_FEATURE_CRYPTO)
+#if defined(__clang__)
+#error "Clang not supported yet"
+#elif defined(__GNUC__)
+#pragma GCC target ("arch=armv8-a+crypto")
```

Comment:
We did this the other way round with the SHA-2 acceleration: when the option is enabled, we check that we've been given the correct compiler flags, rather than silently updating the target from what is specified on the command line

Reply:
With `arch=armv8.2-a+sha3+crypto` and `-O3`,  `eor3` instruction is generated in assembly code of `aesce.c` . That cause `illegal instruction` error.  So I disable `SHA512` in `.travis.yml` .  But that is not expected.

For time being,  I think 
- We should set arch flags per file or per function
- We should not force user enable the compiler flags. Those flags is only used for building library. If user only want to user library, why he must set the flags?

That is not finalized, I am still thinking and testing the issues. I might change above points :)

Reply:
I have posted issues at https://github.com/Mbed-TLS/mbedtls/issues/5387#issuecomment-1379813734

Reply:
I have removed it and followup the method in `sha{256,512}.c`

Reply:
As I've also mentioned [on the architecture document](https://github.com/Mbed-TLS/mbedtls/pull/7004#discussion_r1101958235), we can't just decide to compile for a different target. If the user wants to compile for a particular processor, we can't decide to compile for a different processor!

> And I think we should re-consider the rule due to the illegal instruction  error when runtime detection enabled

I'm sorry, I have no idea 1. what rule you're referring to and 2. what this has to do with an error when compiling some code (?) using a particular compiler option (?).

Reply:
I believe the problem is with runtime detection:

1. You compile in both SHA acceleration (sha3 extension) and AES acceleration (crypto extension) with the intent of shipping it and trusting runtime detection to decide whether it is safe to run the optimised code.
2. The compiler with -O3 option uses an instruction in the AES code that is only present in the sha3 extension but not in the generic crypto extension.
3. On a platform where crypto extension is present but sha3 is not, the runtime detection will pass and run the accelerated AES code which uses an instruction that is not present.

(@yuhaoth please correct me if I am wrong.)

Reply:
@yanesca , Yes, it is the problem. 

And if sha3 extension is present, with sha3 compiler options, we can got best optimization. That's one another point I am thinking about in #7004 .

Reply:
> I'm sorry, I have no idea 1. 
GCC and clang support two method to define target CPU modifier. 
- Option 1 with `pragma`
```c
/* GCC */
#pragma GCC target ("arch=armv8-a+crypto")
/* Clang */
#pragma clang attribute push (__attribute__((target("crypto"))), apply_to=function)
```
- Option2 with attribute
```c
/* GCC */
__attribute__((target("arch=armv8-a+crypto")) void function(void);
/* Clang */
__attribute__((target("crypto")) void function(void);
```

> what rule you're referring to and 2. what this has to do with an error when compiling some code (?) using a particular compiler option (?).

https://github.com/Mbed-TLS/mbedtls/blob/daa65956c3aabf67146a8535cf1ce081bb878dc0/include/mbedtls/check_config.h#L734 and https://github.com/Mbed-TLS/mbedtls/blob/daa65956c3aabf67146a8535cf1ce081bb878dc0/include/mbedtls/check_config.h#L767 . Compiler options are forced . If command line does not include the cpu modifier flags, the place will report error.

I prefer does not force them. And if users need best code size optimization, they should compile application and library with same cpu modifier flags.